node/src/Cardano/Node/API.hs

```diff
@@ -232,102 +239,20 @@ instance Core.HasSlottingVar SettingsCtx where
->
-> And note that this does affect the node and creates actual coupling, so we
-> ought to remove it, eventually.
->
-> I truly believe acid-state or any form of persistent storage is no need
-> however.
-
-Duncan Coutts [1 day ago]
-> I think it's right that no persistence is needed. Certainly in future the
-> node's ledger state will simply report the current on-chain latest version of
-> each application, so Daedalus or the wallet backend just needs to know it's
-> own version and checks occasionally if the on-chain latest version is bigger.
-
--}
-
-{-
--- from Cardano.Wallet.API.Internal.Handlers
-applyUpdate :: PassiveWalletLayer IO -> Handler NoContent
-applyUpdate w = liftIO (WalletLayer.applyUpdate w) >> return NoContent
-
--- from Cardano.WalletLayer.Kernel.hs
--- ...
-        , applyUpdate          = Internal.applyUpdate         w
--- ...
-
--- from Cardano.Wallet.WalletLayer.Kernal.Internal
--- | Apply an update
---
--- NOTE (legacy): 'applyUpdate', "Pos.Wallet.Web.Methods.Misc".
---
--- The legacy implementation does two things:
---
--- 1. Remove the update from the wallet's list of updates
--- 2. Call 'applyLastUpdate' from 'MonadUpdates'
---
--- The latter is implemented in 'applyLastUpdateWebWallet', which literally just
--- a call to 'triggerShutdown'.
---
--- TODO: The other side of the story is 'launchNotifier', where the wallet
--- is /notified/ of updates.
-applyUpdate :: MonadIO m => Kernel.PassiveWallet -> m ()
-applyUpdate w = liftIO $ do
-    update' (w ^. Kernel.wallets) $ RemoveNextUpdate
-    Node.withNodeState (w ^. Kernel.walletNode) $ \_lock -> do
-      doFail <- liftIO $ testLogFInject (w ^. Kernel.walletFInjects) FInjApplyUpdateNoExit
-      unless doFail
-        Node.triggerShutdown
--}
+applyUpdate :: ShutdownContext -> Handler NoContent
+applyUpdate shutdownCtx = liftIO $ do
+    doFail <- testLogFInject (_shdnFInjects shutdownCtx) FInjApplyUpdateNoExit
+    unless doFail (runReaderT triggerShutdown shutdownCtx)
+    pure NoContent
 
-postponeUpdate :: Handler NoContent
-postponeUpdate =
-    throwError err500 { errBody = "This handler is not yet implemented." }
-
-{-
-
--- from Cardano.Wallet.API.Internal.Handlers:
-postponeUpdate :: PassiveWalletLayer IO -> Handler NoContent
-postponeUpdate w = liftIO (WalletLayer.postponeUpdate w) >> return NoContent
-
--- from Cardano.Wallet.WalletLayer.Kernal.Internal
--- | Postpone update
+-- | In the old implementation, we would delete the new update from the
+-- acid-stae database. We no longer persist this information, so postponing an
+-- update is simply a noop.
 --
--- NOTE (legacy): 'postponeUpdate', "Pos.Wallet.Web.Methods.Misc".
-postponeUpdate :: MonadIO m => Kernel.PassiveWallet -> m ()
-postponeUpdate w = update' (w ^. Kernel.wallets) $ RemoveNextUpdate
-
--}
+-- TODO: verify this is a real thought and not, in fact, bad
+postponeUpdate :: Handler NoContent
+postponeUpdate = do
+    pure NoContent
```

Comment:
Yeah, was about to mention it on the `applyUpdate`; we should still be tracking updates in plain memory. Some sort of `MVar` or `IORef` in the node context. So the logic is similar as what we had before, we just don't persist the state.

Reply:
What is the logic, though? From my perspective, the node knows about a new update. When we ask "What's the next update?", then the node responds with the update if available. To update, you just restart the node software. To postpone, you just... don't restart te node software, which is business-as-usual.

Reply:
The current logic is as follows:

- We call `s` the queue of available updates (FIFO), maintained in memory. 
- When a new update is available, the node stores it in an `MVar ConfirmedProposalState` `ucDownloadedUpdate`
- Until now, the wallet was polling this MVar regularly to catch the update and push ii on top of the queue `s`
- When  `GET /api/v1/next-update`, the node will:
    - Clean up `s` by removing all updates that aren't relevant anymore (cf `isActualUpdate` in `Cardano.Wallet.Kernel.DB.Updates`).
    - Peek at the last update on top of `s`
    - If `s` isn't empty, returns the corresponding software version mentioned in the update alongside a 200 response.
    - If `s` is empty, the node returns 404. 

- When `POST /api/v1/postpone-update`, the node remove the oldest update from the the queue `s`. No matter the state of the queue, the node returns a `204 No Content`.

- When `POST  /api/v1/apply-update`, the node remove the oldest update from the queue and no matter the state of the queue, restarts the node with an exit code `20`.

A few interesting things to know (cf `Pos.Network.Update.Download`)

- The last downloaded update is applied, no matter what, when the node restart. The exit code `20` is just a mechanism to tell the underlying launcher that the node should be restarted. But applying the update happens anyway if the node is restarted (after a sigkill, or a clean shutdown). 

- The node doesn't actually download more than one update at a time. 
    - If `ucDownloadedUpdate :: MVar ConfirmedProposalState` hasn't been emptied, it displays a log "We won't download an update for proposal [...]" and ignores the proposal.
    - If `ucDownloadedUpdate` is empty, the node checks whether there's already a file on disk with the update. It throws if there's one. 

---

So basically, there can be only one pending update, and that's the one stored in `ucDownloadedUpdate`. I am afraid that the existing code is sort-of-broken. By emptying this `MVar` early, we are exposing the node to potential crashes... Beside, this makes the underlying queue completely superfluous since there can only be one element in that queue! Perhaps a better way would be to just have another `MVar` telling us whether we've already accounted for the other `ucDownloadedUpdate`, without emptying it. Those `MVar` will be discarded anyway when restarting.